File: 1514-path-with-maximum-probability/1514-path-with-maximum-probability.cpp
class Solution {
public:
    double maxProbability(int n, vector<vector<int>>& edges, vector<double>& succProb, int start, int end) {
        vector<double> dist(n, 0); // min probability will always be 0
        priority_queue<pair<double, int>> pq; // {prob, node} max-heap
        vector<pair<int, double>> adj[n];
        
        // Building the directed graph
        for (int i = 0; i < edges.size(); i++) {
            int u = edges[i][0];
            int v = edges[i][1];
            double prob = succProb[i];
            adj[u].push_back({v, prob});
            adj[v].push_back({u, prob});
        }

        dist[start] = 1.0;
        pq.push({1.0, start});

        while (!pq.empty()) {
            double currProb = pq.top().first;
            int currNode = pq.top().second;
            pq.pop();
            
            // Reached the end node, no need to explore further
            if (currNode == end) break;

            for (auto& it : adj[currNode]) {
                int adjNode = it.first;
                double adjProb = it.second;
                if (currProb * adjProb > dist[adjNode]) {
                    dist[adjNode] = currProb * adjProb;
                    pq.push({dist[adjNode], adjNode});
                }
            }
        }

        return dist[end];
    }
};

File: 1514-path-with-maximum-probability/1514-path-with-maximum-probability.java
class Solution {
    public double maxProbability(int n, int[][] edges, double[] succProb, int start_node, int end_node) {
        List<List<double[]>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) adj.add(new ArrayList<>());

        for (int i = 0; i < edges.length; i++) {
            int u = edges[i][0], v = edges[i][1];
            double p = succProb[i];
            adj.get(u).add(new double[]{v, p});
            adj.get(v).add(new double[]{u, p});
        }

        // Max-heap: highest probability first
        PriorityQueue<double[]> pq = new PriorityQueue<>((a, b) -> Double.compare(b[0], a[0]));
        double[] dist = new double[n];

        dist[start_node] = 1.0;
        pq.offer(new double[]{1.0, start_node});

        while (!pq.isEmpty()) {
            double[] curr = pq.poll();
            double currProb = curr[0];
            int currNode = (int) curr[1];

            for (double[] edge : adj.get(currNode)) {
                int adjNode = (int) edge[0];
                double wt = edge[1];
                if (currProb * wt > dist[adjNode]) {
                    dist[adjNode] = currProb * wt;
                    pq.offer(new double[]{dist[adjNode], adjNode});
                }
            }
        }

        return dist[end_node];
    }
}